### ISSUE TITLE
3 questions regarding the contents of the paper

### ISSUE THREAD

**Issue by hellojinwoo** [RSNirwan/PortfolioAllocationUsingGPLVM#5]
As I am reading your impressive paper 'Applications of GPLVM in Finance', I got to have some questions. It would be mostly due to my lack of knowledge on this topic that hinders understanding your paper, so I will appreciate it a lot if you clarify them.

__All of the questions below are related with the `portfolio_performance.ipynb` file.__

<h3> Q1.  Dimension reduction using GPLVM even for sample covariance and Ledoit Wolf?</h3> 

- Do we need to reduce dimension of ' N assets * D trading days (N * D matrix)' to 'N * Q matrix' (D > Q) before applying the sample covariance and Ledoit Wolf  to optimize a portfolio as well?

- __Where in the codes can I find the line doing the dimension reduction for sample covariance and Ledoit Wolf?__

- Based on the paper, it seems like we need to reduce dimensions of data before optimization because the mathematical equation 18 and 19 on the page no. 6 shows that we use kernels on
&Beta;. Beta is from the matrix whose size is N * Q, according to the page no.5.

- However, when I see codes, it makes me feel puzzled because there seems to be no dimension reduction involved.

```
Y_train = data.loc[now - window:now, :].copy().values.T

.....

if model_name == 'sample_cov': 
    cov = 1./Y_train.shape[1]*Y_train.dot(Y_train.T)
    elbo, r2, lengthscale = np.nan, np.nan, np.nan
    
elif model_name == 'ledoit_wolf': 
    lw = LedoitWolf(store_precision=False, assume_centered=True)
    lw = lw.fit(Y_train.T)
    cov = lw.covariance_
    elbo, r2, lengthscale = np.nan, np.nan, np.nan 
```

- As you can see from the codes above, both of the models do not have the parameters(elbo, r2, lengthscale) needed for dimension reduction.  However, a paper strongly suggests that we need to do dimension reduction for those two portfolio optimization methods as well. Can you kindly specify the codes doing the dimension reduction before those two optimization methods?

<h3> Q2. How often do you rebalance the portfolio? </h3> 

- According to the paper page no. 8, the portfolio is rebalanced every 6 months 

>  Then, we learn _w_ from the past year and buy accordingly for the next six months. Starting from Jan 2008, this procedure is repeated every six months.

- However, the code says that it is rebalanced every 100 days.
```
window = timedelta(200)   # learning period 
delta = timedelta(100)    # predicting period 
tries = 5         

...

start, end = start, end   
rebalancing_times = get_rebalancing_times(start, end, window, delta)
```
- When you type in ` rebalancing_times ` , the result is like ` [Timestamp('2008-07-20 00:00:00'), Timestamp('2008-10-28 00:00:00'), Timestamp('2009-02-05 00:00:00'), Timestamp('2009-05-16 00:00:00'),
... `

- In my best understanding, the first training data is from 2008-01-02 to 2008-07-20, about 6 months, and then based on covaraince matrix from the data from this period, we optimize the portfolio for the first time.

-  __Side question : Which portfolio should we carry for this period (2008-01-02 ~ 2008-07-20) as we do not have data to compute covariance among stocks?__ : Based on the codes, it seems like there is no investment happening during the period. The first column of the variable `df_performance` is `2008-07-21`, 6 months-long trading days after the first day of 2008. 

- And then as delta is 100, it seems that we rebalance the portfolio around every 3-4 moths. Why does the paper say we rebalance the portfolio every 6 months, when delta is 100 days? 

<h3> Q3. What is the name of codes corresponding to the Modern Portfolio Theory? </h3>

- I know this sounds like a stupid question, but please help me figure this out.

```
 if model_name == 'sample_cov': 
        cov = 1./Y_train.shape[1]*Y_train.dot(Y_train.T)
        elbo, r2, lengthscale = np.nan, np.nan, np.nan 
```
- Since sample_cov calculates the covariance function from raw daily_return_rate of stocks matrix, it seems like sample_cov is the traditional way (MPT) of optimizing a portfolio.

- Then what is the code 'linear'? Can you provide me with the portfolio optimization theory name for the method 'linear' (which is the number 0 in model_dict) in your code?

**Comment by rsnirwan** [comment]
Hi Hellojinwoo,

#### Q1: 
* No, we don't need to reduce the dimension of the assets. For portfolio allocation we just need the covariance matrix. 
* The dimensions are not reduces to calculate the sample and ledoit wolf covariance matrix.
Sample covariance is calculated by `cov = 1./Y_train.shape[1]*Y_train.dot(Y_train.T)`
Ledoit wolf:
```
lw = LedoitWolf(store_precision=False, assume_centered=True)
lw = lw.fit(Y_train.T)
cov = lw.covariance_
```
* The GPLVM is a model which reduces the dimension of you data and at the same time calculates the covariance of between the data points. So, either we calculate the sample and the ledoit wolf covariance (done in `if` and `elif`) without the dim reduction or we use the GPLVM the reduce the dimension and get the covariance (the `else` statement. See the `vb()` function and the stan code in GPLVMsInFinance/functions/). 

#### Q2:
* In the paper the portfolio is rebalanced every 6 months. The GitHub code is there to play around with those variables. If you want to rebalance every 6 months and learn from 12 months, set `delta = timedelta(365)/2.` and `window = timedelta(365)`.
* The list `rebalancing_times` shows the dates when we reallocate the resources. So, if the data start at 01.01.2018 with 12 months learning and 6 months prediction, `rebalancing_times` will roughly look like: "[01.01.2009, 01.07.2009, 01.01.2010, ...]". The elements are input to `run_calc()` via `now` and the training matrix is `Y_train = data.loc[now-window:now, :].copy().values.T`.
* There are no assets allocated in the first year, in this case. 

#### Q3:
* Code for eq. 17 (minimal risk portfolio) in the paper is in  GPLVMsInFinance/functions/helper_functions.py (See function `minimal_variance_portfolio()`).
* MPT needs a covariance matrix. It  doesn't matter where it comes from. You can estimate the covariance matrix by the sample covariance or take any other estimator for that. We take GPLVM to estimate the matrix. GPLVM takes a kernel function as input. 'linear' refers to the linear kernel function (eq. 18 in the paper).

**Comment by hellojinwoo** [comment]
Thank you for yorur kind answer.
But I have another question regarding your answer to the question no.1

<h3> Why different Sharpe ratio for simple covariance and Ledoit Wolf with different Q? </h3>

- As you said in your answer, we do not reduce dimension of trading days for calculating covariance using 'sample covaraince' and 'Ledoit Wolf' method.
- However, in the file ` porfolio_performance.ipynb `, these methods with different number of Q have different Sharpe ratio. You can see these different Sharpe ratio in the picture below.
- It is weird because Simple covariance and Ledoit wolf do not use dimension reduction so I cannot understand why these two models with different number of Q have different Sharpe ratio.
- In other words, shouldn't the sharpe ratio be the same for the 'Sample covariance, Q = 3' and 'Sample covariance, Q = 4'  // 'Ledoit wolf, Q =3' and 'Ledoit wolf, Q=4' ?

![another question](https://user-images.githubusercontent.com/34431729/57556925-4bd37400-73b3-11e9-900f-31c7ac7c6470.PNG)

**Comment by rsnirwan** [comment]
There are 2 estimations in the portfolio building case.
* Estimation of covariance matrix K
* Estimation of w_opt (weights for the portfolio, eq. 17)

Even if the estimation of covariance matrix is deterministic (same in each run, e.g. for sample_cov) the estimation of w_opt never is. Due to the constraints on w_opt, eq. 17 has many local minima and each run will give different w_opt and thus slightly different sharpe ratios.
